doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,46 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI sends expected digest with digest type (e.g., ETag)
+2. API checks for existence of ETag in identity table.
+3. if exists, returns key, if not returns an error.
```

Comment:
```suggestion
1. CLI `GET /blobs/` <-- expected digest with digest type (e.g., ETag). API:
    1. Looks for an existing AssetBlob with the given digest(s).
    2. If it exists, return the key (UUID).
    3. It it doesn't exist, return 404.
2. If 404, start upload process.
3. if key, call `POST .../assets/` with the key to register the new asset.
```

Reply:
That is pretty much how it is now -- a responsibility of a good client to explicitly check first. 

I think we should leave above "Check" possibility as is for its own life, and make `/uploads/initialize` with similar request as to check, and do check internally (thus returning e.g. HTTP 409 Conflict with a Location or some other header field to point to existing blob key).  This would preclude "bad" clients to not check first, cause traffic, and us only later returning them a key for an already existing blob at `validate`.  And that is why I liked it when you (I believe) added to have `/uploads/initialize` to get the size and digest.

Reply:
@yarikoptic That would work, but it doesn't resolve the overloading of HTTP actions. I liked the logical separation of having either
`GET /blobs/` -> `POST .../assets/` (no upload involved, so no calls to `/upload/...`)
or
`GET /blobs/` -> `POST /upload/initialize+complete+validate/` -> `POST .../assets/`
That keeps the upload process as a more distinct pathway.

This is just splitting hairs over nitty gritty HTTP design, though. I can implement it your way if you feel strongly.

Reply:
"strongly" would be the wrong word since I am not an experienced API designer ;)  
but it feels that it could help us avoid unnecessary waste of bandwidth so feels "more right" to me.

Note that a client could still do `GET /blobs/` first to proceed current/first path, but going directly to `POST /upload/initialize` (and handling a possible error indicating presence of the blob) would avoid that additional initial `GET /blobs/` call completely

Reply:
How about adding the `GET /blobs/` endpoint so clients have a dedicated, reliable way to get look up existing assets, and making `/upload/initialize/` simply fail when attempting to upload a digest that's already registered?

Reply:
Yes, that is what [I had in mind above as well](https://github.com/dandi/dandi-api/pull/150#discussion_r592744335): `/blobs` is generally good to have, but not needed for "upload logic". Just that `/upload/initialize` should "simply fail" informatively - provide `blobs:key` in the return headers so that client could proceed with it to ```9. CLI calls `POST .../assets/` with the key to register the new asset.```
Note that it does leave possibility for a race condition (GC removes blob:key before CLI calls POST it for a new asset) but I guess it would be very unlikely... may be also there could be some timestamp increment on `blobs` record for that `key` so GC does not even consider blobs which were recently "shown interest in"? Is there something like "access time" for DB records or overhead would be too high?

Reply:
I think we're on the same page, I'm committing this suggestion.

I think race conditions related to the garbage collector are garbage collector problems, which are not in the scope of this design doc.